diff --git a/tensorflow/compiler/xla/BUILD b/tensorflow/compiler/xla/BUILD
index a128eefe51..73cf20b337 100644
--- a/tensorflow/compiler/xla/BUILD
+++ b/tensorflow/compiler/xla/BUILD
@@ -62,6 +62,7 @@ cc_library(
     srcs = [
         "comparison_util.cc",
     ],
+    copts = [ "-ggdb3", "-O0" ],
     hdrs = [
         "comparison_util.h",
     ],
@@ -80,6 +81,7 @@ cc_library(
     srcs = [
         "execution_options_util.cc",
     ],
+    copts = [ "-ggdb3", "-O0" ],
     hdrs = [
         "execution_options_util.h",
     ],
@@ -126,6 +128,7 @@ cc_library(
 cc_library(
     name = "status_macros",
     srcs = ["status_macros.cc"],
+    copts = [ "-ggdb3", "-O0" ],
     hdrs = ["status_macros.h"],
     visibility = ["//visibility:public"],
     deps = [
@@ -175,6 +178,7 @@ cc_library(
 cc_library(
     name = "util",
     srcs = ["util.cc"],
+    copts = [ "-ggdb3", "-O0" ],
     hdrs = [
         "iterator_util.h",
         "map_util.h",
@@ -201,6 +205,7 @@ cc_library(
 cc_library(
     name = "protobuf_util",
     srcs = ["protobuf_util.cc"],
+    copts = [ "-ggdb3", "-O0" ],
     hdrs = [
         "protobuf_util.h",
     ],
@@ -255,6 +260,7 @@ cc_library(
         "shape.h",
         "shape_util.h",
     ],
+    copts = [ "-ggdb3", "-O0" ],
     visibility = ["//visibility:public"],
     deps = [
         ":protobuf_util",
@@ -368,6 +374,7 @@ cc_library(
     name = "literal",
     srcs = ["literal.cc"],
     hdrs = ["literal.h"],
+    copts = [ "-ggdb3", "-O0" ],
     visibility = ["//visibility:public"],
     deps = [
         ":array2d",
@@ -412,6 +419,7 @@ cc_library(
     name = "literal_util",
     srcs = ["literal_util.cc"],
     hdrs = ["literal_util.h"],
+    copts = [ "-ggdb3", "-O0" ],
     visibility = ["//visibility:public"],
     deps = [
         ":array2d",
@@ -440,6 +448,7 @@ cc_library(
     name = "literal_comparison",
     srcs = ["literal_comparison.cc"],
     hdrs = ["literal_comparison.h"],
+    copts = [ "-ggdb3", "-O0" ],
     deps = [
         ":error_spec",
         ":literal",
@@ -456,6 +465,7 @@ cc_library(
     name = "metric_table_report",
     srcs = ["metric_table_report.cc"],
     hdrs = ["metric_table_report.h"],
+    copts = [ "-ggdb3", "-O0" ],
     visibility = ["//visibility:public"],
     deps = [
         ":util",
@@ -573,6 +583,7 @@ cc_library(
     name = "executable_run_options",
     srcs = ["executable_run_options.cc"],
     hdrs = ["executable_run_options.h"],
+    copts = [ "-ggdb3", "-O0" ],
     visibility = ["//visibility:public"],
     deps = [
         ":types",
@@ -583,6 +594,7 @@ cc_library(
     name = "packed_literal_reader",
     srcs = ["packed_literal_reader.cc"],
     hdrs = ["packed_literal_reader.h"],
+    copts = [ "-ggdb3", "-O0" ],
     visibility = [":internal"],
     deps = [
         ":literal",
@@ -617,6 +629,7 @@ cc_library(
     name = "text_literal_reader",
     srcs = ["text_literal_reader.cc"],
     hdrs = ["text_literal_reader.h"],
+    copts = [ "-ggdb3", "-O0" ],
     visibility = [":internal"],
     deps = [
         ":literal",
@@ -653,6 +666,7 @@ cc_library(
     name = "text_literal_writer",
     srcs = ["text_literal_writer.cc"],
     hdrs = ["text_literal_writer.h"],
+    copts = [ "-ggdb3", "-O0" ],
     visibility = [":internal"],
     deps = [
         ":literal",
@@ -715,6 +729,7 @@ cc_library(
     name = "shape_layout",
     srcs = ["shape_layout.cc"],
     hdrs = ["shape_layout.h"],
+    copts = [ "-ggdb3", "-O0" ],
     visibility = ["//visibility:public"],
     deps = [
         ":shape_util",
@@ -729,6 +744,7 @@ cc_library(
     name = "window_util",
     srcs = ["window_util.cc"],
     hdrs = ["window_util.h"],
+    copts = [ "-ggdb3", "-O0" ],
     visibility = ["//visibility:public"],
     deps = [
         ":types",
@@ -744,6 +760,7 @@ cc_library(
     name = "reference_util",
     srcs = ["reference_util.cc"],
     hdrs = ["reference_util.h"],
+    copts = [ "-ggdb3", "-O0" ],
     visibility = ["//visibility:public"],
     deps = [
         ":array2d",
@@ -788,6 +805,7 @@ cc_library(
     name = "sparse_index_array",
     srcs = ["sparse_index_array.cc"],
     hdrs = ["sparse_index_array.h"],
+    copts = [ "-ggdb3", "-O0" ],
     deps = [
         ":array2d",
         ":shape_util",
@@ -859,6 +877,7 @@ cc_library(
     name = "cpu_function_runtime",
     srcs = ["cpu_function_runtime.cc"],
     hdrs = ["cpu_function_runtime.h"],
+    copts = [ "-ggdb3", "-O0" ],
     visibility = [":friends"],
     deps = ["//tensorflow/core:framework_lite"],
 )
diff --git a/tensorflow/compiler/xla/client/BUILD b/tensorflow/compiler/xla/client/BUILD
index b800229bd9..332cde6794 100644
--- a/tensorflow/compiler/xla/client/BUILD
+++ b/tensorflow/compiler/xla/client/BUILD
@@ -27,6 +27,7 @@ cc_library(
     name = "global_data",
     srcs = ["global_data.cc"],
     hdrs = ["global_data.h"],
+    copts = [ "-g", "-O0" ],
     deps = [
         "//tensorflow/compiler/xla:service_interface",
         "//tensorflow/compiler/xla:types",
@@ -42,6 +43,7 @@ cc_library(
     name = "padding",
     srcs = ["padding.cc"],
     hdrs = ["padding.h"],
+    copts = [ "-g", "-O0" ],
     deps = [
         "//tensorflow/compiler/xla:statusor",
         "//tensorflow/compiler/xla:types",
@@ -65,6 +67,7 @@ cc_library(
     name = "client",
     srcs = ["client.cc"],
     hdrs = ["client.h"],
+    copts = [ "-g", "-O0" ],
     deps = [
         ":global_data",
         ":xla_computation",
@@ -90,6 +93,7 @@ cc_library(
     name = "executable_build_options",
     srcs = ["executable_build_options.cc"],
     hdrs = ["executable_build_options.h"],
+    copts = [ "-g", "-O0" ],
     deps = [
         "//tensorflow/compiler/xla:debug_options_flags",
         "//tensorflow/compiler/xla:shape_util",
@@ -107,6 +111,7 @@ cc_library(
     name = "local_client",
     srcs = ["local_client.cc"],
     hdrs = ["local_client.h"],
+    copts = [ "-g", "-O0" ],
     deps = [
         ":client",
         ":executable_build_options",
@@ -193,6 +198,7 @@ cc_library(
     name = "xla_computation",
     srcs = ["xla_computation.cc"],
     hdrs = ["xla_computation.h"],
+    copts = [ "-g", "-O0" ],
     visibility = ["//visibility:public"],
     deps = [
         "//tensorflow/compiler/xla:shape_util",
@@ -208,6 +214,7 @@ cc_library(
     name = "xla_builder",
     srcs = ["xla_builder.cc"],
     hdrs = ["xla_builder.h"],
+    copts = [ "-g", "-O0" ],
     visibility = ["//visibility:public"],
     deps = [
         ":padding",
diff --git a/tensorflow/compiler/xla/service/BUILD b/tensorflow/compiler/xla/service/BUILD
index 1e7a924e35..c795ac3add 100644
--- a/tensorflow/compiler/xla/service/BUILD
+++ b/tensorflow/compiler/xla/service/BUILD
@@ -55,6 +55,7 @@ cc_library(
     name = "bfloat16_support",
     srcs = ["bfloat16_support.cc"],
     hdrs = ["bfloat16_support.h"],
+    copts = ["-ggdb3", "-O0"],
     deps = [
         ":hlo",
     ],
@@ -64,6 +65,7 @@ cc_library(
     name = "bfloat16_conversion_folding",
     srcs = ["bfloat16_conversion_folding.cc"],
     hdrs = ["bfloat16_conversion_folding.h"],
+    copts = ["-ggdb3", "-O0"],
     deps = [
         ":bfloat16_support",
         ":hlo",
@@ -98,6 +100,7 @@ cc_library(
     name = "bfloat16_normalization",
     srcs = ["bfloat16_normalization.cc"],
     hdrs = ["bfloat16_normalization.h"],
+    copts = ["-ggdb3", "-O0"],
     deps = [
         ":bfloat16_support",
         ":hlo",
@@ -135,6 +138,7 @@ cc_library(
     name = "bfloat16_propagation",
     srcs = ["bfloat16_propagation.cc"],
     hdrs = ["bfloat16_propagation.h"],
+    copts = ["-ggdb3", "-O0"],
     deps = [
         ":bfloat16_support",
         ":hlo",
@@ -174,6 +178,7 @@ cc_library(
     name = "dump",
     srcs = ["dump.cc"],
     hdrs = ["dump.h"],
+    copts = ["-ggdb3", "-O0"],
     deps = [
         ":hlo",
         ":hlo_graph_dumper",
@@ -191,6 +196,7 @@ cc_library(
     name = "shape_inference",
     srcs = ["shape_inference.cc"],
     hdrs = ["shape_inference.h"],
+    copts = ["-ggdb3", "-O0"],
     deps = [
         ":hlo",
         "//tensorflow/compiler/xla:shape_util",
@@ -259,6 +265,7 @@ cc_library(
         "hlo_evaluator_typed_visitor_uint64.cc",
         "hlo_evaluator_typed_visitor_uint8.cc",
     ],
+    copts = ["-ggdb3", "-O0"],
     hdrs = ["hlo_evaluator.h"],
     deps = [
         ":dynamic_dimension_inference",
@@ -333,6 +340,7 @@ cc_library(
         "hlo_sharding.cc",
         "hlo_sharding_metadata.cc",
     ],
+    copts = ["-ggdb3", "-O0"],
     hdrs = [
         "dfs_hlo_visitor.h",
         "dfs_hlo_visitor_with_default.h",
@@ -568,6 +576,7 @@ cc_library(
     name = "call_graph",
     srcs = ["call_graph.cc"],
     hdrs = ["call_graph.h"],
+    copts = ["-ggdb3", "-O0"],
     deps = [
         ":hlo",
         "//tensorflow/compiler/xla:status_macros",
@@ -604,6 +613,7 @@ cc_library(
     name = "flatten_call_graph",
     srcs = ["flatten_call_graph.cc"],
     hdrs = ["flatten_call_graph.h"],
+    copts = ["-ggdb3", "-O0"],
     deps = [
         ":call_graph",
         ":hlo",
@@ -674,6 +684,7 @@ cc_library(
     name = "platform_util",
     srcs = ["platform_util.cc"],
     hdrs = ["platform_util.h"],
+    copts = ["-ggdb3", "-O0"],
     deps = [
         ":compiler",
         "//tensorflow/compiler/xla:debug_options_flags",
@@ -691,6 +702,7 @@ cc_library(
     name = "backend",
     srcs = ["backend.cc"],
     hdrs = ["backend.h"],
+    copts = ["-ggdb3", "-O0"],
     deps = [
         ":compiler",
         ":computation_placer",
@@ -716,6 +728,7 @@ cc_library(
     name = "service",
     srcs = ["service.cc"],
     hdrs = ["service.h"],
+    copts = ["-ggdb3", "-O0"],
     deps = [
         ":allocation_tracker",
         ":backend",
@@ -766,6 +779,7 @@ cc_library(
     name = "local_service",
     srcs = ["local_service.cc"],
     hdrs = ["local_service.h"],
+    copts = ["-ggdb3", "-O0"],
     deps = [
         ":backend",
         ":compiler",
@@ -801,6 +815,7 @@ cc_library(
     name = "compile_only_service",
     srcs = ["compile_only_service.cc"],
     hdrs = ["compile_only_service.h"],
+    copts = ["-ggdb3", "-O0"],
     deps = [
         ":backend",
         ":compiler",
diff --git a/tensorflow/compiler/xla/service/local_service.cc b/tensorflow/compiler/xla/service/local_service.cc
index 3470fe5b2c..b964513e1c 100644
--- a/tensorflow/compiler/xla/service/local_service.cc
+++ b/tensorflow/compiler/xla/service/local_service.cc
@@ -42,6 +42,8 @@ limitations under the License.
 #include "tensorflow/core/platform/logging.h"
 #include "tensorflow/core/platform/stream_executor_no_cuda.h"
 
+#define ENDL std::endl << std::flush
+
 namespace xla {
 
 /* static */ StatusOr<std::unique_ptr<LocalService>> LocalService::NewService(
@@ -120,6 +122,7 @@ StatusOr<std::unique_ptr<Executable>> LocalService::CompileExecutable(
     const XlaComputation& computation,
     const absl::Span<const Shape* const> argument_layouts,
     const ExecutableBuildOptions& build_options) {
+  std::cout << "LocalService::CompileExecutable()" << ENDL;
   const HloModuleProto& proto = computation.proto();
   TF_RET_CHECK(proto.has_host_program_shape());
   ProgramShape program_shape(proto.host_program_shape());
diff --git a/tensorflow/compiler/xrt/BUILD b/tensorflow/compiler/xrt/BUILD
index acd984f9e9..f3a48cd102 100644
--- a/tensorflow/compiler/xrt/BUILD
+++ b/tensorflow/compiler/xrt/BUILD
@@ -53,6 +53,7 @@ cc_library(
         "xrt_state.h",
         "xrt_util.h",
     ],
+    copts = [ "-ggdb3", "-O0" ],
     deps = [
         ":xrt_proto",
         "//tensorflow/compiler/jit:xla_device",
diff --git a/tensorflow/core/BUILD b/tensorflow/core/BUILD
index acda23822b..71906c3b50 100644
--- a/tensorflow/core/BUILD
+++ b/tensorflow/core/BUILD
@@ -3240,7 +3240,7 @@ tf_cuda_library(
         "public/version.h",
     ],
     hdrs = CORE_CPU_LIB_HEADERS,
-    copts = tf_copts(),
+    copts = tf_copts() + ["-ggdb3", "-O0"],
     deps = [
         ":bfc_allocator",
         ":graph",
diff --git a/tensorflow/tools/xla_extract/BUILD b/tensorflow/tools/xla_extract/BUILD
index 40cfe8933b..e821a9094f 100644
--- a/tensorflow/tools/xla_extract/BUILD
+++ b/tensorflow/tools/xla_extract/BUILD
@@ -15,6 +15,7 @@ cc_library(
     name = "tf_graph_to_xla_lib",
     srcs = ["tf_graph_to_xla_lib.cc"],
     hdrs = ["tf_graph_to_xla_lib.h"],
+    copts = [ "-g", "-O0" ],
     deps = [
        "//tensorflow/core:framework",
         "//tensorflow/core:lib",
@@ -30,6 +31,7 @@ cc_library(
 tf_cc_binary(
     name = "tf_graph_to_xla",
     srcs = ["tf_graph_to_xla.cc"],
+    copts = [ "-g", "-O0" ],
     visibility = ["//visibility:public"],
     deps = [
         ":tf_graph_to_xla_lib",
